{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import copy\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GA:\n",
    "    def __init__(self):\n",
    "        self.crossover_prob = 0.7\n",
    "        self.mutation_prob = 0.2\n",
    "        self.ring_size = 3\n",
    "        self.buffer = {}\n",
    "    \n",
    "    def _crossover(self, chromosome1, chromosome2):\n",
    "            genes1 = chromosome1[1]\n",
    "            genes2 = chromosome2[1]\n",
    "            genes = np.random.choice(len(genes1), size=2, replace=False)\n",
    "            genes.sort()\n",
    "            child1_genes = genes1[:]\n",
    "            child2_genes = genes2[:]\n",
    "            for i in range(genes[0], genes[1] + 1):\n",
    "                temp_read1 = child1_genes[i][1][:]\n",
    "                temp_read2 = child2_genes[i][1][:]\n",
    "                child1_genes[i] = (genes2[i][0], temp_read2)\n",
    "                child2_genes[i] = (genes1[i][0], temp_read1)\n",
    "            child1 = (0.0, child1_genes)\n",
    "            child2 = (0.0, child2_genes)\n",
    "            return child1, child2\n",
    "\n",
    "    def _mutation(self, population):\n",
    "            score = population[0]\n",
    "            individuals = population[1].copy()\n",
    "            index1, index2 = random.sample(range(len(individuals)), 2)\n",
    "            individuals[index1], individuals[index2] = individuals[index2], individuals[index1]\n",
    "            mutated_individuals = [\n",
    "                (individuo[0], individuo[1]) if i == index1 or i == index2 else individuo\n",
    "                for i, individuo in enumerate(individuals)\n",
    "            ]\n",
    "            mutated_population = (score, mutated_individuals)\n",
    "            return mutated_population\n",
    "    \n",
    "    \n",
    "    # Funzione di fitness, la quale calcola il punteggio per ogni individuo\n",
    "    def _fitness(self, cromosoma):\n",
    "            score = 0\n",
    "            array_valori = [valori for _, valori in cromosoma]\n",
    "\n",
    "            for i in range(len(array_valori) - 1):\n",
    "                finger = array_valori[i]\n",
    "                finger1 = array_valori[i + 1]\n",
    "                # print(\"1 -> \",finger, \"2 -> \",finger1)\n",
    "                # Calcola l'overlap solo se le due sequenze sono identiche\n",
    "\n",
    "                score += self.findOverlapNew(finger, finger1)\n",
    "                # print(\"score\", score)\n",
    "            return score\n",
    "    \n",
    "    \n",
    "    def _evaluatePopulation(self, population, gen):\n",
    "            scores = np.zeros(len(population))\n",
    "            for x in range(len(population)):\n",
    "                scores[x] = self._fitness(population[x])\n",
    "            return scores\n",
    "    \n",
    "    \n",
    "    def findOverlapNew(self, finger, finger1):\n",
    "            overlap = []\n",
    "            min_len = min(len(finger), len(finger1))\n",
    "            for i in range(1, min_len + 1):\n",
    "                if finger[-i:] == finger1[:i]:\n",
    "                    overlap = finger[-i:]\n",
    "            return sum(overlap)\n",
    "    \n",
    "    \n",
    "    # Usata per calcolare L'overlap tra due letture: return overlap\n",
    "    def findOverlapGenoma(self, finger1, finger2):\n",
    "            # print(\"Finger e finger 1\", finger, finger1)\n",
    "            # print(\"Siamo nel metodo findOverlapNew\")\n",
    "            overlap = []\n",
    "            min_len = min(len(finger1), len(finger2))\n",
    "            for i in range(1, min_len + 1):\n",
    "                if finger1[-i:] == finger2[:i]:\n",
    "                    # print(finger[-i:], \"\\n\", finger1[:i])\n",
    "                    overlap = finger1[-i:]\n",
    "                    # print(\"Ancora qui\", overlap)\n",
    "            # print(\"Overlap individuato tra :\",finger, \"\\n\", \"e \", finger1, \"\\n\", \"e di\", sum(overlap))\n",
    "            # print(\"somma:\", sum(overlap))\n",
    "            return overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_ga(self, env, pop, gen, iterazioni):\n",
    "        pop_size = len(pop)\n",
    "        population = pop\n",
    "        pop_fitness = self._evaluatePopulation(population, gen)\n",
    "        print(pop_fitness)\n",
    "        c = np.argmax(pop_fitness)\n",
    "        best_ind = copy.deepcopy(population[c])\n",
    "        best_fit = self._fitness(best_ind)\n",
    "        for generation in range(iterazioni):\n",
    "            print('Iterazione n: ', generation)\n",
    "            mappa_fitness = []\n",
    "            for i in range(len(population)):\n",
    "                individui = population[i]\n",
    "                fitness_corrispondente = pop_fitness[i]\n",
    "                tupla = (fitness_corrispondente, individui)\n",
    "                mappa_fitness.append(tupla)\n",
    "            sorted_data = sorted(mappa_fitness, key=lambda x: x[0], reverse=True)\n",
    "            population = sorted_data\n",
    "            n = len(population) // 2\n",
    "            population = population[:n]\n",
    "            for i in range(0, len(population) - 1, 2):\n",
    "                if np.random.rand() < self.crossover_prob:\n",
    "                    if i < len(population) - 1:\n",
    "                        pop1, pop2 = self._crossover(population[i], population[i + 1])\n",
    "                        population.append(pop1)\n",
    "                        population.append(pop2)\n",
    "            for i in range(len(population)):\n",
    "                if np.random.rand() > self.mutation_prob:\n",
    "                    population[i] = self._mutation(population[i])\n",
    "            population.append(copy.deepcopy((0.0, best_ind)))\n",
    "            population = [individuo for fitness, individuo in population]\n",
    "            pop_fitness = self._evaluatePopulation(population, gen)\n",
    "            print(pop_fitness)\n",
    "            value = pop_fitness.max()\n",
    "            best_ind = population[pop_fitness.argmax()].copy()\n",
    "            if value < best_fit:\n",
    "                best_ind = copy.deepcopy(population[np.argmax(pop_fitness)])\n",
    "                population[0] = copy.deepcopy(best_ind)\n",
    "                best_fit = value\n",
    "            else:\n",
    "                best_ind = population[pop_fitness.argmax()].copy()\n",
    "                best_fit = pop_fitness.max()\n",
    "                indexValue = np.argmax(pop_fitness)\n",
    "                population[0] = copy.deepcopy(population[indexValue])\n",
    "        return best_ind"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Funzioni di utilità"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Definizione stub della funzione qlearning\n",
    "def qlearning(reads, episodes, genome=None, test_each_episode=False):\n",
    "    factor = 1.0 / (len(reads) * max([len(read) for read in reads]))\n",
    "    generations = episodes\n",
    "\n",
    "    _cromosomeInt = []\n",
    "    # print(\"------------\")\n",
    "    print(reads)\n",
    "    sottosequenza_lunghezza = 100\n",
    "\n",
    "    # Caso in cui abbiamo un dataset diviso in piu letture\n",
    "    # readsGenoma = ''.join(reads)\n",
    "    # print(readsGenoma)\n",
    "    i = 0\n",
    "    # print(reads)\n",
    "    genomePartenza = copy.copy(reads)\n",
    "    reads = createDataset(reads, sottosequenza_lunghezza)\n",
    "    # print(reads)\n",
    "    individuo = copy.copy(reads)\n",
    "    # print(len(count_repeats(reads)))\n",
    "\n",
    "    # for x in reads:\n",
    "    # print(\"Lettura: \", x)\n",
    "\n",
    "    # Dimensione del marcatore -> da 4 a 8\n",
    "    countRepeat = 7\n",
    "    dict = count_repeats(reads, countRepeat)\n",
    "    # print(dict)\n",
    "    # print(count_repeats(reads))\n",
    "\n",
    "    # print(\"------------\")\n",
    "    marker = []\n",
    "    # print(\"Le letture sono:\", reads)\n",
    "\n",
    "    #  Marcatori Indipendenti\n",
    "    marksIndependent = 6\n",
    "    max_readss = find_unique_markers(dict, marksIndependent)\n",
    "    # print('3 marcatori distinti', max_readss)\n",
    "\n",
    "    # max_readss = find_unique_markers(dict, 2)\n",
    "    # print('2 marcatori distinti', max_readss)\n",
    "\n",
    "    markers = chiavi = [sequenza for sequenza, valore in max_readss]\n",
    "    # print(\"Combinazione dei marcatori\", markers)\n",
    "\n",
    "    #print(\"------------\")\n",
    "    results = []\n",
    "    results2 = []\n",
    "    for x in range(len(reads)):\n",
    "        # print(\"Lettura\", reads[x])\n",
    "        resul = apply_CFL_to_reads(reads[x], markers)\n",
    "        # risp = apply_CFL_to_reads(reads[x], markers)\n",
    "        # print(\"Due marcatori:\", risp)\n",
    "        # print(\"Tre marcatori:\", resul)\n",
    "        results.append(resul)\n",
    "        # results2.append(risp)\n",
    "\n",
    "    # print(\"Risultato ottenuto:\", results)\n",
    "\n",
    "    _intA = compute_fingerprint_by_list_factors(results)\n",
    "    # _intB = compute_fingerprint_by_list_factors(results2)\n",
    "    # print(\"------------\")\n",
    "\n",
    "    # Popolazione\n",
    "    num_ind = 1500\n",
    "    iterazioni = 1500\n",
    "    ga = GA()\n",
    "\n",
    "    # Creo una lista di indici\n",
    "    indices = list(range(len(_intA)))\n",
    "    # print(_intA)\n",
    "    indConfront = copy.copy(_intA)\n",
    "    popolazioni_mescolate = generate_random_populations(_intA, num_ind, len(_intA))\n",
    "    # popolazioni_mescolate2 = generate_random_populations(_intB, num_ind, len(_intB))\n",
    "    # print(popolazioni_mescolate)\n",
    "\n",
    "    # print(\"----\")\n",
    "    for i, sublist in enumerate(popolazioni_mescolate):\n",
    "        #print(sublist)\n",
    "        chiavi = [tupla[0] for tupla in sublist]\n",
    "        array_di_interi = [tupla[1] for tupla in sublist]\n",
    "        # Conversione della lista di chiavi in una singola stringa\n",
    "        stringa_chiavi = ''.join(chiavi)\n",
    "        chunks = [stringa_chiavi[i:i + len(individuo[0])] for i in range(0, len(stringa_chiavi), len(individuo[0]))]\n",
    "        # print(chunks)\n",
    "        # if individuo == chunks:\n",
    "        # print(\"Trovato\", chunks)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_random_populations(data, num_ind, length):\n",
    "    # implementation of the function goes here\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_zeros(array):\n",
    "    no_zeros = [element for element in array if element != 0]\n",
    "    return no_zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CFL(word, T):\n",
    "    CFL_list = []\n",
    "    k = 0\n",
    "    while k < len(word):\n",
    "        i = k + 1\n",
    "        j = k + 2\n",
    "        while True:\n",
    "            if j == len(word) + 1 or word[j - 1] < word[i - 1]:\n",
    "                while k < i:\n",
    "                    CFL_list.append(word[k:k + j - i])\n",
    "                    k = k + j - i\n",
    "                break\n",
    "            else:\n",
    "                if word[j - 1] > word[i - 1]:\n",
    "                    i = k + 1\n",
    "                else:\n",
    "                    i = i + 1\n",
    "                j = j + 1\n",
    "    return CFL_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_repeats(reads, countRepeat):\n",
    "    repeats_count = {}\n",
    "    max_length = len(reads[0])\n",
    "    for k in range(countRepeat, max_length + 1):\n",
    "        for read in reads:\n",
    "            prefix = read[:k]\n",
    "            repeats_count.setdefault(prefix, 0)\n",
    "            repeats_count[prefix] += sum(1 for r in reads if prefix in r)\n",
    "    return repeats_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_CFL_to_reads(reads, markers):\n",
    "    cfl_list = []\n",
    "    marker_indices = []\n",
    "\n",
    "    # Trova tutti gli indici dei marcatori nella sequenza\n",
    "    start = 0\n",
    "    while start < len(reads):\n",
    "        found_indices = [(reads.find(marker, start), marker) for marker in markers if reads.find(marker, start) != -1]\n",
    "        if not found_indices:\n",
    "            break\n",
    "        idx, marker = min(found_indices, key=lambda x: x[0])\n",
    "        marker_indices.append((idx, marker))\n",
    "        start = idx + len(marker)\n",
    "\n",
    "    if not marker_indices:\n",
    "        cfl_list.append(reads)\n",
    "        segments = CFL(reads[:], None)\n",
    "        result = (reads, segments)\n",
    "        return result\n",
    "\n",
    "    # Se la sequenza inizia senza un marcatore, aggiungi la parte iniziale come CFL\n",
    "    if marker_indices and marker_indices[0][0] > 0:\n",
    "        cfl_list.append(CFL(reads[:marker_indices[0][0]], None))\n",
    "\n",
    "    # Calcola le CFL sulla base degli indici dei marcatori\n",
    "    i = 0\n",
    "    while i < len(marker_indices) - 1:\n",
    "        start_idx, current_marker = marker_indices[i]\n",
    "        next_idx, next_marker = marker_indices[i + 1]\n",
    "\n",
    "        # Controlla se i marcatori non sono consecutivi\n",
    "        if start_idx + len(current_marker) <= next_idx:\n",
    "            cfl_list.append(CFL(reads[start_idx:next_idx], None))\n",
    "\n",
    "        else:\n",
    "            # Trova il prossimo marcatore che non si sovrappone\n",
    "            j = i + 1\n",
    "            while j < len(marker_indices) and start_idx + len(current_marker) > marker_indices[j][0]:\n",
    "                j += 1\n",
    "            if j < len(marker_indices):\n",
    "                cfl_list.append(CFL(reads[start_idx:marker_indices[j][0]], None))\n",
    "            else:\n",
    "                cfl_list.append(CFL(reads[start_idx:], None))\n",
    "                break\n",
    "        i += 1\n",
    "\n",
    "    # Aggiungi l'ultimo segmento se esiste\n",
    "    if marker_indices:\n",
    "        last_idx, last_marker = marker_indices[-1]\n",
    "        if last_idx + len(last_marker) <= len(reads):\n",
    "            cfl_list.append(CFL(reads[last_idx:], None))\n",
    "\n",
    "    lista_appiattita = [elemento for sottolista in cfl_list for elemento in sottolista]\n",
    "    result = (reads, lista_appiattita)\n",
    "    # print(\"Lettura: \", reads, \"\\nMarcatori Utilizzati: \", markers, \"\\nRappresentazione CFL: \", cfl_list)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_fingerprint_by_list_factors(original_list):\n",
    "    oggetto_trasformato = []\n",
    "    for chiave, lista_stringhe in original_list:\n",
    "        # Converti ogni stringa nella lista nella sua lunghezza\n",
    "        lunghezze = [len(stringa) for stringa in lista_stringhe]\n",
    "        oggetto_trasformato.append((chiave, lunghezze))\n",
    "    return oggetto_trasformato"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_unique_markers(reads, num_markers):\n",
    "    # Ordina le sequenze per frequenza e lunghezza in ordine decrescente\n",
    "    sorted_reads = sorted(reads.items(), key=lambda x: (-x[1], -len(x[0])))\n",
    "\n",
    "    unique_markers = []\n",
    "\n",
    "    for current_marker in sorted_reads:\n",
    "        # Verifica che il marcatore corrente non sia un sottoinsieme di nessuno dei marcatori selezionati\n",
    "        is_subset = False\n",
    "        for selected_marker in unique_markers:\n",
    "            if current_marker[0] in selected_marker[0] or selected_marker[0] in current_marker[0]:\n",
    "                is_subset = True\n",
    "                break\n",
    "        if not is_subset:\n",
    "            unique_markers.append(current_marker)\n",
    "        if len(unique_markers) == num_markers:\n",
    "            break\n",
    "\n",
    "    return unique_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_random_populations(reads, num_populations, population_size):\n",
    "    populations = [reads]\n",
    "    for _ in range(num_populations - 1):\n",
    "        population = random.sample(reads, population_size)\n",
    "        populations.append(population)\n",
    "    return populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assemble_genome_with_overlaps(reads):\n",
    "    # print(\"reads:\", reads)\n",
    "    genome = reads[0][0]\n",
    "    # print(\"qui\", genome)\n",
    "    for i in range(len(reads) - 1):\n",
    "        current_read, current_overlaps = reads[i]\n",
    "        current_read2, current_overlaps2 = reads[i + 1]\n",
    "        # print(current_read, current_overlaps, \"\\n\", current_read2, current_overlaps2)\n",
    "\n",
    "        lista = GA.findOverlapGenoma(None, current_overlaps, current_overlaps2)\n",
    "        # print(\"lista\", lista)\n",
    "        if (len(lista) > 0):\n",
    "            stringa_concatenata = current_read2[sum(lista):]\n",
    "            # print(\"Stringa conc\", stringa_concatenata)\n",
    "            genome += stringa_concatenata\n",
    "            # print(\"Genoma\", genome)\n",
    "        else:\n",
    "            genome += current_read2\n",
    "            # print(\"Genoma\", genome)\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def levenshtein(s, t, costs=(1, 1, 1)):\n",
    "    \"\"\"\n",
    "                iterative_levenshtein(s, t) -> ldist\n",
    "                ldist is the Levenshtein distance between the strings\n",
    "                s and t.\n",
    "                For all i and j, dist[i,j] will contain the Levenshtein\n",
    "                distance between the first i characters of s and the\n",
    "                first j characters of t\n",
    "\n",
    "                costs: a tuple or a list with three integers (d, i, s)\n",
    "                       where d defines the costs for a deletion\n",
    "                             i defines the costs for an insertion and\n",
    "                             s defines the costs for a substitution\n",
    "            \"\"\"\n",
    "    rows = len(s) + 1\n",
    "    cols = len(t) + 1\n",
    "    deletes, inserts, substitutes = costs\n",
    "\n",
    "    dist = [[0 for x in range(cols)] for x in range(rows)]\n",
    "    # source prefixes can be transformed into empty strings\n",
    "    # by deletions:\n",
    "    for row in range(1, rows):\n",
    "        dist[row][0] = row * deletes\n",
    "\n",
    "    # target prefixes can be created from an empty source string\n",
    "    # by inserting the characters\n",
    "    for col in range(1, cols):\n",
    "        dist[0][col] = col * inserts\n",
    "\n",
    "    for col in range(1, cols):\n",
    "        for row in range(1, rows):\n",
    "            if s[row - 1] == t[col - 1]:\n",
    "                cost = 0\n",
    "            else:\n",
    "                cost = substitutes\n",
    "            dist[row][col] = min(dist[row - 1][col] + deletes,\n",
    "                                 dist[row][col - 1] + inserts,\n",
    "                                 dist[row - 1][col - 1] + cost)  # substitution\n",
    "    return dist[rows - 1][cols - 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createDataset(dataset, lunghezza_sottosequenza):\n",
    "    # Definiamo i 4 caratteri\n",
    "    # caratteri = ['A', 'C', 'G','T']\n",
    "\n",
    "    # Creiamo un dataset di 2000 caratteri scelti casualmente tra i 4\n",
    "    # dataset = ''.join(np.random.choice(caratteri) for _ in range(2000))\n",
    "\n",
    "    # print(dataset)\n",
    "    # Dividiamo il dataset in sottosequenze di 100 caratteri\n",
    "    # sottosequenze = [dataset[i:i + 100] for i in range(0, len(dataset), 100)]\n",
    "    # return sottosequenze\n",
    "\n",
    "    # Calcola il numero massimo di pezzi sovrapposti\n",
    "    sottosequenze = []\n",
    "    indice_inizio = 0\n",
    "    k = 0\n",
    "    while indice_inizio < len(dataset) - lunghezza_sottosequenza + 1:\n",
    "        # Genera la sottosequenza corrente\n",
    "        sottosequenza = dataset[indice_inizio:indice_inizio + lunghezza_sottosequenza]\n",
    "        # print(\"sottosequenza estratta\", sottosequenza)\n",
    "        sottosequenze.append(sottosequenza)\n",
    "        # print(\"sottosequenza estratta\", sottosequenze)\n",
    "        # Seleziona un indice casuale per l'overlap\n",
    "        indice_casuale = random.randint(1, lunghezza_sottosequenza - 1)\n",
    "        # print(\"INDICE CASUALE\", indice_casuale)\n",
    "        # Aggiorna l'indice di inizio per la prossima sottosequenza\n",
    "        indice_inizio += indice_casuale\n",
    "\n",
    "        # Verifica se la fine della sequenza è stata raggiunta\n",
    "        if indice_inizio + lunghezza_sottosequenza > len(dataset):\n",
    "            # Aggiungi l'ultima sottosequenza che include la fine della sequenza\n",
    "            sottosequenze.append(dataset[-lunghezza_sottosequenza:])\n",
    "            break\n",
    "        k = k + 1\n",
    "    \"\"\"\n",
    "    i = 0\n",
    "    car = 0\n",
    "    print(sottosequenze)\n",
    "    for x in sottosequenze:\n",
    "        print(len(x))\n",
    "        car += len(x)\n",
    "        i += 1\n",
    "\n",
    "    print(\"DIM\", len(sottosequenze),i, car)\n",
    "    \"\"\"\n",
    "\n",
    "    return sottosequenze"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
